docs/tutorials/Containers.md

```diff
@@ -178,27 +190,48 @@ To run Singularity on a job scheduler, the singularity command needs to be passe
 For example, in SLURM, we can use the normal SLURM configuration as explained in the [SLURM documentation](../backends/SLURM), however we'll add a `submit-docker` block to execute when a task is tagged with a docker container. 
 
 When constructing this block, there are a few things to keep in mind:
-- Make sure Singularity is loaded (and in our path), for example you can call `module load Singularity/3.0.1`.
-- We should treat worker nodes as if they do not have stable access to the internet or build access, so we will pull the container before execution of the task.
-- It's a good idea to ask Singularity to build the image into the execution directory of the task as an artifact, and to save rendering time on the worker node.
-- We'll use the docker name (potentially including the docker hash) to generate a compliant filename that can be looked up later.
-- If the container exists, there is no need to rebuild the container.
+- Make sure Singularity is loaded (and in PATH). If `module` is installed for 
+  example you can call `module load Singularity`. If the cluster admin has made
+  a Singularity module available. Alternatively you can alter the `PATH` 
+  variable directly or simply use `/path/to/singularity` 
+  directly in the config.
+- We should treat worker nodes as if they do not have stable access to the 
+  internet or build access, so we will pull the container before the task is 
+  submit to the cluster.
+- It's a good idea to use a Singularity cache so that same images should only
+  have to be pulled once. Make sure you set the `SINGULARITY_CACHEDIR` 
+  environment variable to a location on the filesystem that is reachable by the
+  worker nodes!
+- If we are using a cache we need to ensure that submit processes started by
+  Cromwell do not pull to the same cache at the same time. This may corrupt the
+  cache. We can prevent this by implementing a filelock with `flock` and 
+  pulling the image before the job is submitted. The flock and pull command 
+  needs to be placed *before* the submit command so all pull commands are 
+  executed on the same node. This is necessary for the filelock to work.
+- As mentioned above the `--containall` flag is **important** for 
+  reproducibility.
 
 ```
 submit-docker = """
-    module load Singularity/3.0.1
-  
-    # Build the Docker image into a singularity image, using the head node
-    DOCKER_NAME=$(sed -e 's/[^A-Za-z0-9._-]/_/g' <<< ${docker})
-    IMAGE=${cwd}/$DOCKER_NAME.sif
-    if [ ! -f $IMAGE ]; then
-        singularity pull $IMAGE docker://${docker}
+    # Make sure the SINGULARITY_CACHEDIR variable is set. If not use a default
+    # based on the users home.
+    if [ -z $SINGULARITY_CACHEDIR ]; 
+        then CACHE_DIR=$HOME/.singularity/cache
+        else CACHE_DIR=$SINGULARITY_CACHEDIR
     fi
-  
+    # Make sure cache dir exists so lock file can be created by flock
+    mkdir -p $CACHE_DIR  
+    LOCK_FILE=$CACHE_DIR/singularity_pull_flock
```

Comment:
Not sure whether it's relevant (as I don't know if these pulls will run across many nodes), but just wanted to raise that `flock` based approaches for synchronization such as here can come unstuck when locking isn't cluster consistent.

It's not unheard of for people to be running Lustre mounted `localflock` so locking is only node local etc.